prereise/gather/demanddata/transportation_electrification/generate_scaling_factors.py

```diff
@@ -273,3 +269,104 @@ def get_efs_vmt_projection_for_state(efs_annual_data, electrification_scenario=N
     }
 
     return efs_for_state
+
+
+def generate_scaling_factor(efs_vmt_for_state, vmt_for_ua_perc, vmt_for_ra_perc):
+    """Calculate for each year the scaling factors by vehicle technology in uraban and
+    rural areas.
+
+    :param dict efs_vmt_for_state: keys are state abbreviations and values are data
+        frames enclosing the projected annual annual VMT by vehicle technology and year
+        as returned by :func:`get_efs_vmt_projection_for_state`.
+    :param dict vmt_for_ua_perc: keys are state abbreviations and values series giving
+        VMT for urban areas in state. THis is returned by
+        :func:`calculate_urban_rural_fraction`
+    :param dict vmt_for_ua_perc: same as ``vmt_for_ua_perc`` but for rural areas.
+    :return: (*tuple*) -- dictionary for uraban and rural area. Keys are years, values
+        are data frames enclosing scaling factors for each vehicle type.
+    """
+    old2new = {
+        "ELECTRIC LIGHT-DUTY AUTO - 100 MILE RANGE": "LDV Car - 100 mi",
+        "ELECTRIC LIGHT-DUTY AUTO - 200 MILE RANGE": "LDV Car - 200 mi",
+        "ELECTRIC LIGHT-DUTY AUTO - 300 MILE RANGE": "LDV Car - 300 mi",
+        "ELECTRIC LIGHT-DUTY TRUCK - 100 MILE RANGE": "LDV Truck - 100 mi",
+        "ELECTRIC LIGHT-DUTY TRUCK - 200 MILE RANGE": "LDV Truck - 200 mi",
+        "ELECTRIC LIGHT-DUTY TRUCK - 300 MILE RANGE": "LDV Truck - 300 mi",
+        "BATTERY ELECTRIC MEDIUM-DUTY VEHICLE": "MDV Truck",
+        "ELECTRIC HEAVY DUTY VEHICLE": "HDV Truck",
+    }
+
+    rural_scaling_factor = {}
+    urban_scaling_factor = {}
+    for s in state2abv.values():
+        efs_value = efs_vmt_for_state[s].pivot(
+            index="YEAR", columns="DEMAND_TECHNOLOGY", values="VALUE"
+        )
+        for y in efs_value.index:
+            rural_scaling_factor[y] = pd.concat(
+                [
+                    rural_scaling_factor.get(y, pd.DataFrame()),
+                    pd.DataFrame(
+                        (efs_value.loc[y] * vmt_for_ra_perc[s].values).to_dict(),
+                        index=[s],
+                    )
+                    .rename(old2new, axis=1)
+                    .rename_axis("State"),
+                ]
+            )
+            urban_scaling_factor[y] = pd.concat(
+                [
+                    urban_scaling_factor.get(y, pd.DataFrame()),
+                    pd.concat(
+                        [
+                            pd.DataFrame(
+                                {
+                                    "state": s,
+                                    "UA": vmt_for_ua_perc[s]
+                                    .index.str.split(",")
+                                    .str[0],
+                                },
+                                index=vmt_for_ua_perc[s].index,
+                            ),
+                            pd.DataFrame(
+                                {i: j for i, j in efs_value.loc[y].to_dict().items()},
```

Comment:
Not sure what we want to achieve here, the resultant dict seems to be equivalent to `efs_value.loc[y].to_dict()`, or all we need is simply `efs_value.loc[y].to_frame(name=vmt_for_ua_perc[s].index).T`

Reply:
It works for rural (implemented in last commit) but fails for urban because the shape of `efs_value` is (1,8) and there can be more than one urban area per state. See below where it fails when doing `to_frame(name=vmt_for_ua_perc[s].index)`:
```
raceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/brdo/CEM/PreREISE/prereise/gather/demanddata/transportation_electrification/generate_scaling_factors.py", line 334, in generate_scaling_factor
    efs_value.loc[y]
  File "/Users/brdo/.local/share/virtualenvs/PreREISE-Yr21Ys7H/lib/python3.10/site-packages/pandas/core/ops/__init__.py", line 436, in f
    self, other = align_method_FRAME(self, other, axis, flex=True, level=level)
  File "/Users/brdo/.local/share/virtualenvs/PreREISE-Yr21Ys7H/lib/python3.10/site-packages/pandas/core/ops/__init__.py", line 264, in align_method_FRAME
    raise ValueError(
ValueError: Unable to coerce to DataFrame, shape must be (1, 8): given (2, 1)
```

Reply:
I see. Then we could at least do following without recreating the dict:
```python
pd.DataFrame(efs_value.loc[y].to_dict(), index=vmt_for_ua_perc[s].index)
```

Reply:
Indeed. Done